src/io/reader.rs

```diff
@@ -0,0 +1,108 @@
+use std::fs::File;
+use std::io::{self, BufRead, BufReader, Cursor, Read, Seek, SeekFrom};
+use std::path::Path;
+
+use dynimage::DynamicImage;
+use image::ImageFormat;
+use {ImageError, ImageResult};
+
+use super::free_functions;
+
+/// A multi-format image reader.
+///
+/// Wraps an input reader to facilitate automatic detection of an image's format, appropriate
+/// decoding method, and dispatches into the set of supported `ImageDecoder` implementations.
+pub struct Reader<R: Read> {
+    /// The reader.
+    inner: R,
+    /// The format, if one has been set or deduced.
+    format: Option<ImageFormat>,
+}
+
+impl<R: Read> Reader<R> {
+    pub fn new(reader: R) -> Self {
+        Reader {
+            inner: reader,
+            format: None,
+        }
+    }
+
+    /// Construct a reader with specified format.
+    pub fn with_format(reader: R, format: ImageFormat) -> Self {
+        Reader {
+            inner: reader,
+            format: Some(format),
+        }
+    }
+
+    /// Get the currently determined format.
+    pub fn format(&self) -> Option<ImageFormat> {
+        self.format
+    }
+
+    /// Unwrap the reader.
+    pub fn into_inner(self) -> R {
+        self.inner
+    }
+}
+
+impl Reader<BufReader<File>> {
+    /// Open a file to read. Format will be guessed.
+    /// Note: Failure to guess is not an error. The `Err` is only for `File::open`.
+    pub fn open<P>(path: P) -> io::Result<Self> where P: AsRef<Path> {
```

Comment:
Perhaps call this `from_path`?

Reply:
The most important part of this operation is, in my eyes, `File::open`. I suppose the answer depends on whether the file content should be inspect for magic bytes. The name `from_path` suggests to me that this is not the case, we could also provide both.

Reply:
Maybe `Reader::open` should not already read parts of the file, so that it alludes to `File::open` more closely. Instead, we may want the file magic method to appear under another name if we can find one that is sufficiently distinct from `open`. This would clear up the confusion as well and seems slightly cleaner to me.

-----------

I could have sworn that I submitted that comment earlier but it doesn't appear so.

Reply:
A very good argument for `open` only inspecting the filename would the the current `image::open` function only doing that.

This means we need a better name for the function that *does* inspect the file. Not sure if `read` (similar to `fs::read`) fits that functionality correctly but it is at least a candidate.

Reply:
Not withstanding the already open PR to change that behaviour, #933 of course.

Reply:
There is already a `guess_format` function, so I think it is fine to have `open only consider the path.

Reply:
It's not quite the same, but that may be fixable on its own.

* The current `open` guess the format, and if that doesn't return anything falls back to path.

* `Reader::from_path(..)?.guess_format()?` would always overwrite the guess from the path.

Maybe it should simply not overwrite the current format in `guess_format` when the guess comes back empty.

Reply:
It's what the documentation says already, just my implementation being wrong.

Reply:
Can `guess_format` ever be wrong? Specifically, if it returns some format (rather than "unknown"), it is ever possible that the image could be decoded as something else? If not, I'd be in favor of something like `Reader::from_path(..)?.guess_format()?` being the default method.

Reply:
Yes, if that behaviour is unwanted. Specifically there are a variety of programs who *only* rely on paths for formats (particularly on Windows but also `graphicx` in Latex) and where it would be surprising if a broken image gets parsed with some other format and leads to unexpected pixel types.

It's maybe a too cautious but the combination of operations is really straightforward.

Reply:
Does that even make sense? One could also `clear_format` afterwards if it is unwanted.

Reply:
Guessing the format could certainly break user expectations, but I was more wondering if there are any "[polyglot](https://en.wikipedia.org/wiki/Polyglot_(computing))" images that could be say both a valid PNG and a valid JPEG? That particular pair isn't possible because any PNG starts with `89 50 4E 47` and any valid JPEG starts with `FF D8`, but perhaps there are some pairs where that could happen?

The reason I ask is that it leads to a clear priority between filenames and magic bytes: if the user asks us to consider both, then we always want to use what the magic bytes say first and only fall back to file extensions if there aren't any recognizable magic bytes.

Reply:
I don't think there are any based on the existing table of magic bytes. But some (`ico`: `\x00\x00\x01\x00`) are certainly not very unique overall files.